src/opnsense/mvc/app/models/OPNsense/IPsec/IPsec.php

```diff
@@ -0,0 +1,188 @@
+     * @param $nodeKey string Fully-qualified key of the keyPair instance within a model
+     * @param $keyPair \OPNsense\Base\FieldTypes\BaseField Field instance of a keyPair
+     * @param $messages \Phalcon\Validation\Message\Group Validation message group
+     */
+    private function validateKeyPair($nodeKey, $keyPair, $messages)
+    {
+        $publicKey = $privateKey = null;
+        if (empty((string)$keyPair->keyType))
+            return;
+
+        // Validate public key
+        if (!empty((string)$keyPair->publicKey)) {
+            try {
+                $publicKey = $this->parseCryptographicKey(
+                    (string)$keyPair->publicKey,
+                    (string)$keyPair->keyType . '-public'
+                );
+            } catch (\Exception $e) {
+                $messages->appendMessage(new \Phalcon\Validation\Message($e->getMessage(), $nodeKey . '.publicKey'));
+            }
+        }
+
+        // Validate private key
+        if (!empty((string)$keyPair->privateKey)) {
+            try {
+                $privateKey = $this->parseCryptographicKey(
+                    (string)$keyPair->privateKey,
+                    (string)$keyPair->keyType . '-private'
+                );
+            } catch (\Exception $e) {
+                $messages->appendMessage(new \Phalcon\Validation\Message($e->getMessage(), $nodeKey . '.privateKey'));
+            }
+        }
+
+        // Compare SHA1 fingerprint of public and private keys to check if they belong to each other
+        if ($publicKey && $privateKey) {
+            if ($publicKey['fingerprint'] !== $privateKey['fingerprint']) {
+                $messages->appendMessage(new \Phalcon\Validation\Message(
+                    gettext('This private key does not belong to the given public key.'), $nodeKey . '.privateKey'
+                ));
+            }
+        }
+
+        // Store sanitized representation of keys and cache key statistics
+        $keyPair->publicKey = $publicKey ? $publicKey['pem'] : (string)$keyPair->publicKey;
+        $keyPair->privateKey = $privateKey ? $privateKey['pem'] : (string)$keyPair->privateKey;
+        $keyPair->keySize = $publicKey ? $publicKey['size'] : 0;
+        $keyPair->keyFingerprint = $publicKey ? $publicKey['fingerprint'] : '';
+    }
+
+    /**
+     * Parse a cryptographic key of a given type using OpenSSL and return an array of informational data.
+     * @param $keyString string
+     * @param $keyType string
+     * @return array
+     */
+    public function parseCryptographicKey($keyString, $keyType)
+    {
+        // Attempt to load key with correct type
+        if ($keyType === 'rsa-public') {
+            $key = openssl_pkey_get_public($keyString);
+        } elseif ($keyType === 'rsa-private') {
+            $key = openssl_pkey_get_private($keyString);
+        } else {
+            throw new \InvalidArgumentException('Unsupported key type: ' . $keyType);
```

Comment:
since the other arguments are passed through `gettext()` I would expect something similar here.

Reply:
Updated, this exception now also makes use of the translation system. Good catch!